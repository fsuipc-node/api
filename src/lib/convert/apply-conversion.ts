import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
