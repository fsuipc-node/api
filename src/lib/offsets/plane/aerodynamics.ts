import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  incidenceAlpha: new Offset({
    value: 0x2ED0,
    name: 'incidenceAlpha',
    description: 'incidence alpha (body AoA) - radians',
    category: OffsetCategory.AERODYNAMICS,
    type: Type.Double,
    permission: 'r',
  }),
};
