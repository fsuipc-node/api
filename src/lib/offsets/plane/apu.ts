import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  APUGeneratorSwitch: new Offset({
    value: 0xB51,
    name: 'APUGeneratorSwitch',
    category: OffsetCategory.APU,
    description: 'APU generator switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  APUGeneratorActive: new Offset({
    value: 0xB52,
    name: 'APUGeneratorActive',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUOnFire: new Offset({
    value: 0xB53,
    name: 'APUOnFire',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUMaxRPM: new Offset({
    value: 0xB54,
    name: 'APUMaxRPM',
    category: OffsetCategory.APU,
    description: 'APU generator max RPM percent',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'r',
  }),
  APUStarter: new Offset({
    value: 0xB58,
    name: 'APUStarter',
    category: OffsetCategory.APU,
    description: 'APU starter',
    convert: '{VAL} ? 1 : 0',
    type: Type.Single,
    permission: 'rw',
  }),
  APUVoltage: new Offset({
    value: 0xB5C,
    name: 'APUVoltage',
    category: OffsetCategory.APU,
    description: 'APU generator voltage',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'rw',
  }),
};
