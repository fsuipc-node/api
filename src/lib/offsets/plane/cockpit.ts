import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const cockpit: OffsetList = {
  navLights: new Offset({
    value: 0x280,
    name: 'navLights',
    category: OffsetCategory.COCKPIT,
    description: 'nav lights',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  strobeLights: new Offset({
    value: 0x281,
    name: 'strobeLights',
    category: OffsetCategory.COCKPIT,
    description: 'strobe lights',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  landingLights: new Offset({
    value: 0x28C,
    name: 'landingLights',
    category: OffsetCategory.COCKPIT,
    description: 'landing lights',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'rw',
  }),
  lights: new Offset({
    value: 0x0D0C,
    name: 'lights',
    category: OffsetCategory.COCKPIT,
    description: 'all lights - FS2000+',
    type: Type.BitArray,
    convert: 'lightsMapping',
    mapping: true,
    length: 2,
    permission: 'rw',
  }),
  alternateStaticAirSource: new Offset({
    value: 0x29B,
    name: 'alternateStaticAirSource',
    category: OffsetCategory.COCKPIT,
    description: 'Alternate static air source selected',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  pitotHeat: new Offset({
    value: 0x29C,
    name: 'pitotHeat',
    category: OffsetCategory.COCKPIT,
    description: 'pitot heat active',
    convert: '!!{VAL}',
    type: Type.Int16,
    permission: 'rw',
  }),
  altimeterSettings: new Offset({
    value: 0x330,
    name: 'altimeterSettings',
    category: OffsetCategory.COCKPIT,
    description: 'altimeters settings - mb',
    convert: '{VAL} / 16',
    type: Type.UInt16,
    permission: 'rw',
  }),
  altimeterSettingsG1000: new Offset({
    value: 0x332,
    name: 'altimeterSettingsG1000',
    category: OffsetCategory.COCKPIT,
    description: 'G1000 altimeters settings - mb',
    convert: '{VAL} / 16',
    type: Type.UInt16,
    permission: 'rw',
  }),
  turnCoordinatorPosition: new Offset({
    value: 0x36E,
    name: 'turnCoordinatorPosition',
    category: OffsetCategory.COCKPIT,
    description: 'Turn coordinator ball position - + to the right, - to the left, 0 balanced',
    type: Type.SByte,
    permission: 'r',
  }),
  turnRate: new Offset({
    value: 0x37C,
    name: 'turnRate',
    category: OffsetCategory.COCKPIT,
    description: 'Turn rate needle - minutes (- to the left, + to the right)',
    type: Type.Int16,
    convert: 'Math.round({VAL} / 1024)',
    permission: 'r',
  }),
  preciseTurnCoordinatorPosition: new Offset({
    value: 0x380,
    name: 'preciseTurnCoordinatorPosition',
    category: OffsetCategory.COCKPIT,
    description: 'Turn coordinator ball position - + to the right, - to the left, 0 balanced',
    type: Type.Single,
    permission: 'r',
  }),
  preciseTurnRate: new Offset({
    value: 0x384,
    name: 'preciseTurnRate',
    category: OffsetCategory.COCKPIT,
    description: 'Turn rate needle - minutes (- to the left, + to the right)',
    type: Type.Single,
    permission: 'r',
  }),
  stallWarning: new Offset({
    value: 0x36C,
    name: 'stallWarning',
    category: OffsetCategory.COCKPIT,
    description: 'stall warning',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  overspeedWarning: new Offset({
    value: 0x36D,
    name: 'overspeedWarning',
    category: OffsetCategory.COCKPIT,
    description: 'overspeed warning',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  displayIAS: new Offset({
    value: 0x612,
    name: 'displayIAS',
    category: OffsetCategory.COCKPIT,
    description: 'display IAS - <= FS2000',
    type: Type.UInt16,
    permission: 'rw',
  }),
  suctionPressure: new Offset({
    value: 0xB18,
    name: 'suctionPressure',
    category: OffsetCategory.COCKPIT,
    description: 'gyro suction - inHg',
    type: Type.Double,
    permission: 'r',
  }),
  bleedAirSourceControl: new Offset({
    value: 0xB50,
    name: 'bleedAirSourceControl',
    category: OffsetCategory.COCKPIT,
    description: '0=Min, 1=auto, 2=Off, 3=APU, 4=Engines - FSX A321: 0=Auto, 1=Shut (off), 2=APU, 3=Engines',
    type: Type.Byte,
    permission: 'rw',
  }),
  gyroDrift: new Offset({
    value: 0xC3E,
    name: 'gyroDrift',
    category: OffsetCategory.COCKPIT,
    description: 'gyro drift - degrees',
    convert: 'Math.round({VAL} * 360 / 65536)',
    type: Type.Int16,
    permission: 'rw',
  }),
  NDScale738A321EFIS: new Offset({
    value: 0xE00,
    name: 'NDScale738A321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 738 and A321 EFIS: ND scale: 738: 0=5nm up to 7=640nm - A321: 0=10nm up to 5=320nm',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDMode738EFIS: new Offset({
    value: 0xE02,
    name: 'NDMode738EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 738 EFIS: ND mode:0=APP, 1=VOR, 2=MAP',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDMapItems738A321EFIS: new Offset({
    value: 0xE04,
    name: 'NDMapItems738A321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 738 and A321 EFIS: ND map items shown: 738: 0=WPT, 1=APT, 2=NDB, 3=VOR - A321: 0=WPT, 1=VOR, 2=NDB, 3=APT',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDVORADF1Switch738EFIS: new Offset({
    value: 0xE06,
    name: 'NDVORADF1Switch738EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 738 EFIS: ND VOR/ADF1 switch: 0=VOR, 1=OFF, 2=ADF',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDVORADF2Switch738EFIS: new Offset({
    value: 0xE08,
    name: 'NDVORADF2Switch738EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 738 EFIS: ND VOR/ADF2 switch: 0=VOR, 1=OFF, 2=ADF',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDArcCentered738EFIS: new Offset({
    value: 0xE0A,
    name: 'NDArcCentered738EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 738 EFIS: ND arc=0, centred=1',
    type: Type.UInt16,
    permission: 'rw',
  }),
  autopilotSpeedSwitch738EFIS: new Offset({
    value: 0xE0C,
    name: 'autopilotSpeedSwitch738EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 738 EFIS: AP speed/mach C/O button (pressed if 1, not pressed if 0). ',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  NDModeA321EFIS: new Offset({
    value: 0xE0E,
    name: 'NDModeA321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default A321 EFIS: ND mode: 0=ILS, 1=VOR, 2=NAV, 3=ARC',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDVORADF1SwitchA321EFIS: new Offset({
    value: 0xE10,
    name: 'NDVORADF1SwitchA321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default A321 EFIS: ND VOR/ADF1 switch: 0=VOR, 1=OFF, 2=ADF',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDVORADF2SwitchA321EFIS: new Offset({
    value: 0xE12,
    name: 'NDVORADF2SwitchA321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default A321 EFIS: ND VOR/ADF2 switch: 0=VOR, 1=OFF, 2=ADF',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDInHghPASwitchA321EFIS: new Offset({
    value: 0xE14,
    name: 'NDInHghPASwitchA321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default A321 EFIS: ND InHg/hPA switch, 0=InHg, 1=hPA',
    type: Type.UInt16,
    permission: 'rw',
  }),
  NDILSSwitchA321EFIS: new Offset({
    value: 0xE16,
    name: 'NDILSSwitchA321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default A321 EFIS: ND ILS mode button, 0 = off, 1=on',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'rw',
  }),
  autopilotSpeedSwitch321EFIS: new Offset({
    value: 0xE18,
    name: 'autopilotSpeedSwitch321EFIS',
    category: OffsetCategory.COCKPIT,
    description: 'Default 321 EFIS: AP speed/mach C/O button (pressed if 1, not pressed if 0). ',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  altitudeChangeRate321EFIS: new Offset({
    value: 0xE1A,
    name: 'altitudeChangeRate',
    category: OffsetCategory.COCKPIT,
    description: 'Default A321 EFIS: Altitude change rate switch (0 = 100, 1=1000)',
    convert: '{VAL} === 0 ? 100 : 1000',
    type: Type.UInt16,
    permission: 'r',
  }),
  fuelFlowAtCruise: new Offset({
    value: 0x1270,
    name: 'fuelFlowAtCruise',
    category: OffsetCategory.COCKPIT,
    description: 'Estimated fuel flow at cruise - lb/h',
    type: Type.Int32,
    permission: 'r',
  }),
  gyroCompassHeading: new Offset({
    value: 0x2B00,
    name: 'gyroCompassHeading',
    category: OffsetCategory.COCKPIT,
    description: 'gyro compass heading (MAG) including drift',
    convert: 'Math.round({VAL})',
    type: Type.Double,
    permission: 'r',
  }),
  panelAutoFeatherArmSwitch: new Offset({
    value: 0x2E88,
    name: 'panelAutoFeatherArmSwitch',
    category: OffsetCategory.COCKPIT,
    description: 'panel auto-feather arm switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  standbyVacuumCircuitActive: new Offset({
    value: 0x2E90,
    name: 'standbyVacuumCircuitActive',
    category: OffsetCategory.COCKPIT,
    description: 'standby vacuum circuit active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  propSyncActive: new Offset({
    value: 0x2EC8,
    name: 'propSyncActive',
    category: OffsetCategory.COCKPIT,
    description: 'prop sync active',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  barberPoleMach: new Offset({
    value: 0x2F50,
    name: 'barberPoleMach',
    category: OffsetCategory.COCKPIT,
    description: 'barber pole mach',
    type: Type.Double,
    permission: 'r',
  }),
  attitudeCageButton: new Offset({
    value: 0x2F68,
    name: 'attitudeCageButton',
    category: OffsetCategory.COCKPIT,
    description: 'attitude cage button',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  attitudeIndicatorPitch: new Offset({
    value: 0x2F70,
    name: 'attitudeIndicatorPitch',
    category: OffsetCategory.COCKPIT,
    description: 'attitude indicator pitch - degrees',
    type: Type.Double,
    permission: 'r',
  }),
  attitudeIndicatorBank: new Offset({
    value: 0x2F78,
    name: 'attitudeIndicatorBank',
    category: OffsetCategory.COCKPIT,
    description: 'attitude indicator bank - degrees',
    type: Type.Double,
    permission: 'r',
  }),
};
