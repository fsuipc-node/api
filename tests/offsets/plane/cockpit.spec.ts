import { cockpit as offsets } from '@offsets/plane/cockpit';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/cockpit', () => {
  const offsetsTestCases = [
    { name: 'navLights', value: 0, expectedResult: false },
    { name: 'navLights', value: 1, expectedResult: true },
    { name: 'strobeLights', value: 0, expectedResult: false },
    { name: 'strobeLights', value: 1, expectedResult: true },
    { name: 'landingLights', value: 0, expectedResult: false },
    { name: 'landingLights', value: 1, expectedResult: true },
    { name: 'alternateStaticAirSource', value: 0, expectedResult: false },
    { name: 'alternateStaticAirSource', value: 1, expectedResult: true },
    { name: 'pitotHeat', value: 0, expectedResult: false },
    { name: 'pitotHeat', value: 1, expectedResult: true },
    { name: 'altimeterSettings', value: 32, expectedResult: 2 },
    { name: 'altimeterSettingsG1000', value: 32, expectedResult: 2 },
    { name: 'turnRate', value: 2765, expectedResult: 3 },
    { name: 'turnRate', value: 2055, expectedResult: 2 },
    { name: 'stallWarning', value: 0, expectedResult: false },
    { name: 'stallWarning', value: 1, expectedResult: true },
    { name: 'overspeedWarning', value: 0, expectedResult: false },
    { name: 'overspeedWarning', value: 1, expectedResult: true },
    { name: 'gyroDrift', value: 9102, expectedResult: 50 },
    { name: 'autopilotSpeedSwitch738EFIS', value: 0, expectedResult: false },
    { name: 'autopilotSpeedSwitch738EFIS', value: 1, expectedResult: true },
    { name: 'NDILSSwitchA321EFIS', value: 0, expectedResult: false },
    { name: 'NDILSSwitchA321EFIS', value: 1, expectedResult: true },
    { name: 'autopilotSpeedSwitch321EFIS', value: 0, expectedResult: false },
    { name: 'autopilotSpeedSwitch321EFIS', value: 1, expectedResult: true },
    { name: 'altitudeChangeRate321EFIS', value: 0, expectedResult: 100 },
    { name: 'altitudeChangeRate321EFIS', value: 1, expectedResult: 1000 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
