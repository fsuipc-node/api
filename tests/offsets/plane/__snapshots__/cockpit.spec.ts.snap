// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/cockpit offsets list should have required properties 1`] = `
Object {
  "NDArcCentered738EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default 738 EFIS: ND arc=0, centred=1",
    "length": undefined,
    "mapping": undefined,
    "name": "NDArcCentered738EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3594,
  },
  "NDILSSwitchA321EFIS": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "Default A321 EFIS: ND ILS mode button, 0 = off, 1=on",
    "length": undefined,
    "mapping": undefined,
    "name": "NDILSSwitchA321EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3606,
  },
  "NDInHghPASwitchA321EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default A321 EFIS: ND InHg/hPA switch, 0=InHg, 1=hPA",
    "length": undefined,
    "mapping": undefined,
    "name": "NDInHghPASwitchA321EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3604,
  },
  "NDMapItems738A321EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default 738 and A321 EFIS: ND map items shown: 738: 0=WPT, 1=APT, 2=NDB, 3=VOR - A321: 0=WPT, 1=VOR, 2=NDB, 3=APT",
    "length": undefined,
    "mapping": undefined,
    "name": "NDMapItems738A321EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3588,
  },
  "NDMode738EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default 738 EFIS: ND mode:0=APP, 1=VOR, 2=MAP",
    "length": undefined,
    "mapping": undefined,
    "name": "NDMode738EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3586,
  },
  "NDModeA321EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default A321 EFIS: ND mode: 0=ILS, 1=VOR, 2=NAV, 3=ARC",
    "length": undefined,
    "mapping": undefined,
    "name": "NDModeA321EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3598,
  },
  "NDScale738A321EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default 738 and A321 EFIS: ND scale: 738: 0=5nm up to 7=640nm - A321: 0=10nm up to 5=320nm",
    "length": undefined,
    "mapping": undefined,
    "name": "NDScale738A321EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3584,
  },
  "NDVORADF1Switch738EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default 738 EFIS: ND VOR/ADF1 switch: 0=VOR, 1=OFF, 2=ADF",
    "length": undefined,
    "mapping": undefined,
    "name": "NDVORADF1Switch738EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3590,
  },
  "NDVORADF1SwitchA321EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default A321 EFIS: ND VOR/ADF1 switch: 0=VOR, 1=OFF, 2=ADF",
    "length": undefined,
    "mapping": undefined,
    "name": "NDVORADF1SwitchA321EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3600,
  },
  "NDVORADF2Switch738EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default 738 EFIS: ND VOR/ADF2 switch: 0=VOR, 1=OFF, 2=ADF",
    "length": undefined,
    "mapping": undefined,
    "name": "NDVORADF2Switch738EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3592,
  },
  "NDVORADF2SwitchA321EFIS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Default A321 EFIS: ND VOR/ADF2 switch: 0=VOR, 1=OFF, 2=ADF",
    "length": undefined,
    "mapping": undefined,
    "name": "NDVORADF2SwitchA321EFIS",
    "permission": "rw",
    "type": 5,
    "value": 3602,
  },
  "alternateStaticAirSource": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "Alternate static air source selected",
    "length": undefined,
    "mapping": undefined,
    "name": "alternateStaticAirSource",
    "permission": "rw",
    "type": 0,
    "value": 667,
  },
  "altimeterSettings": Offset {
    "category": "cockpit",
    "convert": "{VAL} / 16",
    "description": "altimeters settings - mb",
    "length": undefined,
    "mapping": undefined,
    "name": "altimeterSettings",
    "permission": "rw",
    "type": 5,
    "value": 816,
  },
  "altimeterSettingsG1000": Offset {
    "category": "cockpit",
    "convert": "{VAL} / 16",
    "description": "G1000 altimeters settings - mb",
    "length": undefined,
    "mapping": undefined,
    "name": "altimeterSettingsG1000",
    "permission": "rw",
    "type": 5,
    "value": 818,
  },
  "altitudeChangeRate321EFIS": Offset {
    "category": "cockpit",
    "convert": "{VAL} === 0 ? 100 : 1000",
    "description": "Default A321 EFIS: Altitude change rate switch (0 = 100, 1=1000)",
    "length": undefined,
    "mapping": undefined,
    "name": "altitudeChangeRate",
    "permission": "r",
    "type": 5,
    "value": 3610,
  },
  "attitudeCageButton": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "attitude cage button",
    "length": undefined,
    "mapping": undefined,
    "name": "attitudeCageButton",
    "permission": "rw",
    "type": 6,
    "value": 12136,
  },
  "attitudeIndicatorBank": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "attitude indicator bank - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "attitudeIndicatorBank",
    "permission": "r",
    "type": 8,
    "value": 12152,
  },
  "attitudeIndicatorPitch": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "attitude indicator pitch - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "attitudeIndicatorPitch",
    "permission": "r",
    "type": 8,
    "value": 12144,
  },
  "autopilotSpeedSwitch321EFIS": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "Default 321 EFIS: AP speed/mach C/O button (pressed if 1, not pressed if 0). ",
    "length": undefined,
    "mapping": undefined,
    "name": "autopilotSpeedSwitch321EFIS",
    "permission": "r",
    "type": 5,
    "value": 3608,
  },
  "autopilotSpeedSwitch738EFIS": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "Default 738 EFIS: AP speed/mach C/O button (pressed if 1, not pressed if 0). ",
    "length": undefined,
    "mapping": undefined,
    "name": "autopilotSpeedSwitch738EFIS",
    "permission": "r",
    "type": 5,
    "value": 3596,
  },
  "barberPoleMach": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "barber pole mach",
    "length": undefined,
    "mapping": undefined,
    "name": "barberPoleMach",
    "permission": "r",
    "type": 8,
    "value": 12112,
  },
  "bleedAirSourceControl": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "0=Min, 1=auto, 2=Off, 3=APU, 4=Engines - FSX A321: 0=Auto, 1=Shut (off), 2=APU, 3=Engines",
    "length": undefined,
    "mapping": undefined,
    "name": "bleedAirSourceControl",
    "permission": "rw",
    "type": 0,
    "value": 2896,
  },
  "displayIAS": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "display IAS - <= FS2000",
    "length": undefined,
    "mapping": undefined,
    "name": "displayIAS",
    "permission": "rw",
    "type": 5,
    "value": 1554,
  },
  "fuelFlowAtCruise": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Estimated fuel flow at cruise - lb/h",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelFlowAtCruise",
    "permission": "r",
    "type": 3,
    "value": 4720,
  },
  "gyroCompassHeading": Offset {
    "category": "cockpit",
    "convert": "Math.round({VAL})",
    "description": "gyro compass heading (MAG) including drift",
    "length": undefined,
    "mapping": undefined,
    "name": "gyroCompassHeading",
    "permission": "r",
    "type": 8,
    "value": 11008,
  },
  "gyroDrift": Offset {
    "category": "cockpit",
    "convert": "Math.round({VAL} * 360 / 65536)",
    "description": "gyro drift - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "gyroDrift",
    "permission": "rw",
    "type": 2,
    "value": 3134,
  },
  "landingLights": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "landing lights",
    "length": undefined,
    "mapping": undefined,
    "name": "landingLights",
    "permission": "rw",
    "type": 5,
    "value": 652,
  },
  "lights": Offset {
    "category": "cockpit",
    "convert": "lightsMapping",
    "description": "all lights - FS2000+",
    "length": 2,
    "mapping": true,
    "name": "lights",
    "permission": "rw",
    "type": 12,
    "value": 3340,
  },
  "navLights": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "nav lights",
    "length": undefined,
    "mapping": undefined,
    "name": "navLights",
    "permission": "rw",
    "type": 0,
    "value": 640,
  },
  "overspeedWarning": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "overspeed warning",
    "length": undefined,
    "mapping": undefined,
    "name": "overspeedWarning",
    "permission": "r",
    "type": 0,
    "value": 877,
  },
  "panelAutoFeatherArmSwitch": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "panel auto-feather arm switch",
    "length": undefined,
    "mapping": undefined,
    "name": "panelAutoFeatherArmSwitch",
    "permission": "rw",
    "type": 6,
    "value": 11912,
  },
  "pitotHeat": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "pitot heat active",
    "length": undefined,
    "mapping": undefined,
    "name": "pitotHeat",
    "permission": "rw",
    "type": 2,
    "value": 668,
  },
  "preciseTurnCoordinatorPosition": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Turn coordinator ball position - + to the right, - to the left, 0 balanced",
    "length": undefined,
    "mapping": undefined,
    "name": "preciseTurnCoordinatorPosition",
    "permission": "r",
    "type": 9,
    "value": 896,
  },
  "preciseTurnRate": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Turn rate needle - minutes (- to the left, + to the right)",
    "length": undefined,
    "mapping": undefined,
    "name": "preciseTurnRate",
    "permission": "r",
    "type": 9,
    "value": 900,
  },
  "propSyncActive": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "prop sync active",
    "length": undefined,
    "mapping": undefined,
    "name": "propSyncActive",
    "permission": "rw",
    "type": 6,
    "value": 11976,
  },
  "stallWarning": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "stall warning",
    "length": undefined,
    "mapping": undefined,
    "name": "stallWarning",
    "permission": "r",
    "type": 0,
    "value": 876,
  },
  "standbyVacuumCircuitActive": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "standby vacuum circuit active",
    "length": undefined,
    "mapping": undefined,
    "name": "standbyVacuumCircuitActive",
    "permission": "rw",
    "type": 6,
    "value": 11920,
  },
  "strobeLights": Offset {
    "category": "cockpit",
    "convert": "!!{VAL}",
    "description": "strobe lights",
    "length": undefined,
    "mapping": undefined,
    "name": "strobeLights",
    "permission": "rw",
    "type": 0,
    "value": 641,
  },
  "suctionPressure": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "gyro suction - inHg",
    "length": undefined,
    "mapping": undefined,
    "name": "suctionPressure",
    "permission": "r",
    "type": 8,
    "value": 2840,
  },
  "turnCoordinatorPosition": Offset {
    "category": "cockpit",
    "convert": undefined,
    "description": "Turn coordinator ball position - + to the right, - to the left, 0 balanced",
    "length": undefined,
    "mapping": undefined,
    "name": "turnCoordinatorPosition",
    "permission": "r",
    "type": 1,
    "value": 878,
  },
  "turnRate": Offset {
    "category": "cockpit",
    "convert": "Math.round({VAL} / 1024)",
    "description": "Turn rate needle - minutes (- to the left, + to the right)",
    "length": undefined,
    "mapping": undefined,
    "name": "turnRate",
    "permission": "r",
    "type": 2,
    "value": 892,
  },
}
`;
